File: bindings/ocaml/lib/safe_email.ml
(* SPDX-License-Identifier: PMPL-1.0 *)
(* SPDX-FileCopyrightText: 2025 Hyperpolymath *)

(** Safe email validation and parsing operations. *)

type email_parts = {
  local_part: string;
  domain: string;
}

let is_valid email =
  match String.split_on_char '@' email with
  | [local_part; domain] ->
      String.length local_part > 0 &&
      String.length domain >= 3 &&
      String.contains domain '.' &&
      domain.[0] <> '.' &&
      domain.[String.length domain - 1] <> '.'
  | _ -> false

let split email =
  if not (is_valid email) then None
  else
    match String.split_on_char '@' email with
    | [local_part; domain] -> Some { local_part; domain }
    | _ -> None

let get_domain email =
  match split email with
  | Some parts -> Some parts.domain
  | None -> None

let get_local_part email =
  match split email with
  | Some parts -> Some parts.local_part
  | None -> None

let normalize email =
  match split email with
  | Some parts -> Some (parts.local_part ^ "@" ^ String.lowercase_ascii parts.domain)
  | None -> None

File: bindings/julia/src/SafeEmail.jl
# SPDX-License-Identifier: PMPL-1.0
# SPDX-FileCopyrightText: 2025 Hyperpolymath

"""
    SafeEmail

Safe email validation and parsing operations.
"""
module SafeEmail

export EmailParts, is_valid, split_email, get_domain, get_local_part, normalize

"""
    EmailParts

Represents the parts of an email address.
"""
struct EmailParts
    local_part::String
    domain::String
end

"""
    is_valid(email::AbstractString) -> Bool

Check if an email address is valid (basic check).
"""
function is_valid(email::AbstractString)::Bool
    parts = split(email, "@")
    length(parts) != 2 && return false

    local_part, domain = parts
    length(local_part) == 0 && return false
    length(domain) < 3 && return false
    !contains(domain, ".") && return false
    startswith(domain, ".") && return false
    endswith(domain, ".") && return false

    true
end

"""
    split_email(email::AbstractString) -> Union{EmailParts, Nothing}

Split an email into local part and domain.
"""
function split_email(email::AbstractString)::Union{EmailParts, Nothing}
    !is_valid(email) && return nothing

    parts = split(email, "@")
    EmailParts(String(parts[1]), String(parts[2]))
end

"""
    get_domain(email::AbstractString) -> Union{String, Nothing}

Extract the domain from an email address.
"""
function get_domain(email::AbstractString)::Union{String, Nothing}
    parts = split_email(email)
    parts === nothing ? nothing : parts.domain
end

"""
    get_local_part(email::AbstractString) -> Union{String, Nothing}

Extract the local part from an email address.
"""
function get_local_part(email::AbstractString)::Union{String, Nothing}
    parts = split_email(email)
    parts === nothing ? nothing : parts.local_part
end

"""
    normalize(email::AbstractString) -> Union{String, Nothing}

Normalize an email address (lowercase domain).
"""
function normalize(email::AbstractString)::Union{String, Nothing}
    parts = split_email(email)
    parts === nothing && return nothing
    parts.local_part * "@" * lowercase(parts.domain)
end

end # module
